# change winner/loser rank to home/away rank

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_absolute_percentage_error

from sklearn.ensemble import AdaBoostRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score

In [2]:
df = pd.read_csv('wk12.csv')

In [3]:
df

Wk       Date      Time  Day            Winner  Pts Unnamed: 6  \
0     1  27-Aug-22   7:00 PM  Sat  Florida Atlantic   43        NaN   
1     1  27-Aug-22   5:00 PM  Sat     Florida State   47        NaN   
2     1  27-Aug-22   4:00 PM  Sat          Illinois   38        NaN   
3     1  27-Aug-22  10:00 PM  Sat            Nevada   23          @   
4     1  27-Aug-22   3:30 PM  Sat  Nevada-Las Vegas   52        NaN   
..   ..        ...       ...  ...               ...  ...        ...   
644  11   5-Nov-22   7:30 PM  Sat         (12) Utah   45        NaN   
645  11   5-Nov-22   3:30 PM  Sat        Utah State   27        NaN   
646  11   5-Nov-22   3:30 PM  Sat  Washington State   52          @   
647  11   5-Nov-22  12:00 PM  Sat  Western Kentucky   59          @   
648  11   5-Nov-22  12:00 PM  Sat         Wisconsin   23        NaN   

                Loser  Pts.1  
0           Charlotte     13  
1            Duquesne      7  
2             Wyoming      6  
3    New Mexico State     12  
4         Idaho State     21  
..                ...    ...  
644           Arizona     20  
645        New Mexico     10  
646          Stanford     14  
647         Charlotte      7  
648          Maryland     10  

[649 rows x 9 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 649 entries, 0 to 648
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Wk          649 non-null    int64 
 1   Date        649 non-null    object
 2   Time        647 non-null    object
 3   Day         649 non-null    object
 4   Winner      649 non-null    object
 5   Pts         649 non-null    int64 
 6   Unnamed: 6  229 non-null    object
 7   Loser       649 non-null    object
 8   Pts.1       649 non-null    int64 
dtypes: int64(3), object(6)
memory usage: 45.8+ KB


In [5]:
location = df['Unnamed: 6']
home = []

for index, row in df.iterrows():
    location = (row['Unnamed: 6'])
    if (location == 'N'):
        home.append('Neutral')
    elif (location == '@'):
        home.append(row['Loser'])
    else:
        home.append(row['Winner'])



In [6]:
df['home_team'] = home

In [7]:
away = []
for index, row in df.iterrows():
    if (row['home_team'] == row['Winner']):
        away.append(row['Loser'])
    elif (row['home_team'] == row['Loser']):
        away.append(row['Winner'])
    else:
        away.append('neutral')

In [8]:
df['away'] = away

In [9]:
df

Wk       Date      Time  Day            Winner  Pts Unnamed: 6  \
0     1  27-Aug-22   7:00 PM  Sat  Florida Atlantic   43        NaN   
1     1  27-Aug-22   5:00 PM  Sat     Florida State   47        NaN   
2     1  27-Aug-22   4:00 PM  Sat          Illinois   38        NaN   
3     1  27-Aug-22  10:00 PM  Sat            Nevada   23          @   
4     1  27-Aug-22   3:30 PM  Sat  Nevada-Las Vegas   52        NaN   
..   ..        ...       ...  ...               ...  ...        ...   
644  11   5-Nov-22   7:30 PM  Sat         (12) Utah   45        NaN   
645  11   5-Nov-22   3:30 PM  Sat        Utah State   27        NaN   
646  11   5-Nov-22   3:30 PM  Sat  Washington State   52          @   
647  11   5-Nov-22  12:00 PM  Sat  Western Kentucky   59          @   
648  11   5-Nov-22  12:00 PM  Sat         Wisconsin   23        NaN   

                Loser  Pts.1         home_team              away  
0           Charlotte     13  Florida Atlantic         Charlotte  
1            Duquesne      7     Florida State          Duquesne  
2             Wyoming      6          Illinois           Wyoming  
3    New Mexico State     12  New Mexico State            Nevada  
4         Idaho State     21  Nevada-Las Vegas       Idaho State  
..                ...    ...               ...               ...  
644           Arizona     20         (12) Utah           Arizona  
645        New Mexico     10        Utah State        New Mexico  
646          Stanford     14          Stanford  Washington State  
647         Charlotte      7         Charlotte  Western Kentucky  
648          Maryland     10         Wisconsin          Maryland  

[649 rows x 11 columns]

In [10]:
df['Winner'].value_counts()

Coastal Carolina             8
San Jose State               7
Troy                         7
South Alabama                7
Texas-San Antonio            7
                            ..
(13) North Carolina State    1
(24) Texas A&M               1
(21) Texas                   1
(15) Tennessee               1
(10) UCLA                    1
Name: Winner, Length: 243, dtype: int64

In [11]:
home = df['home_team']

home_rank = []

for i in home:
    if (i[2] == ")"):
        home_rank.append(i[1:2])
        if(i[1] == "("):
            home_rank.append(i[1:3])
    else:
        home_rank.append(99)

In [12]:
away = df['away']

away_rank = []

for i in away:
    if (i[2] == ")"):
        away_rank.append(i[1:2])
        if(i[1] == "("):
            away_rank.append(i[1:3])
    else:
        away_rank.append(99)

In [13]:
df['home_rank'] = home_rank

In [14]:
df['away_rank'] = away_rank

In [15]:
df

Wk       Date      Time  Day            Winner  Pts Unnamed: 6  \
0     1  27-Aug-22   7:00 PM  Sat  Florida Atlantic   43        NaN   
1     1  27-Aug-22   5:00 PM  Sat     Florida State   47        NaN   
2     1  27-Aug-22   4:00 PM  Sat          Illinois   38        NaN   
3     1  27-Aug-22  10:00 PM  Sat            Nevada   23          @   
4     1  27-Aug-22   3:30 PM  Sat  Nevada-Las Vegas   52        NaN   
..   ..        ...       ...  ...               ...  ...        ...   
644  11   5-Nov-22   7:30 PM  Sat         (12) Utah   45        NaN   
645  11   5-Nov-22   3:30 PM  Sat        Utah State   27        NaN   
646  11   5-Nov-22   3:30 PM  Sat  Washington State   52          @   
647  11   5-Nov-22  12:00 PM  Sat  Western Kentucky   59          @   
648  11   5-Nov-22  12:00 PM  Sat         Wisconsin   23        NaN   

                Loser  Pts.1         home_team              away home_rank  \
0           Charlotte     13  Florida Atlantic         Charlotte        99   
1            Duquesne      7     Florida State          Duquesne        99   
2             Wyoming      6          Illinois           Wyoming        99   
3    New Mexico State     12  New Mexico State            Nevada        99   
4         Idaho State     21  Nevada-Las Vegas       Idaho State        99   
..                ...    ...               ...               ...       ...   
644           Arizona     20         (12) Utah           Arizona        99   
645        New Mexico     10        Utah State        New Mexico        99   
646          Stanford     14          Stanford  Washington State        99   
647         Charlotte      7         Charlotte  Western Kentucky        99   
648          Maryland     10         Wisconsin          Maryland        99   

    away_rank  
0          99  
1          99  
2          99  
3          99  
4          99  
..        ...  
644        99  
645        99  
646        99  
647        99  
648        99  

[649 rows x 13 columns]

In [16]:
df["home_team_code"] = df["home_team"].astype("category").cat.codes

In [17]:
df["away_team_code"] = df["away"].astype("category").cat.codes

In [18]:
homes = df['home_team']

In [19]:
df1 = df.copy()

In [20]:
df1

Wk       Date      Time  Day            Winner  Pts Unnamed: 6  \
0     1  27-Aug-22   7:00 PM  Sat  Florida Atlantic   43        NaN   
1     1  27-Aug-22   5:00 PM  Sat     Florida State   47        NaN   
2     1  27-Aug-22   4:00 PM  Sat          Illinois   38        NaN   
3     1  27-Aug-22  10:00 PM  Sat            Nevada   23          @   
4     1  27-Aug-22   3:30 PM  Sat  Nevada-Las Vegas   52        NaN   
..   ..        ...       ...  ...               ...  ...        ...   
644  11   5-Nov-22   7:30 PM  Sat         (12) Utah   45        NaN   
645  11   5-Nov-22   3:30 PM  Sat        Utah State   27        NaN   
646  11   5-Nov-22   3:30 PM  Sat  Washington State   52          @   
647  11   5-Nov-22  12:00 PM  Sat  Western Kentucky   59          @   
648  11   5-Nov-22  12:00 PM  Sat         Wisconsin   23        NaN   

                Loser  Pts.1         home_team              away home_rank  \
0           Charlotte     13  Florida Atlantic         Charlotte        99   
1            Duquesne      7     Florida State          Duquesne        99   
2             Wyoming      6          Illinois           Wyoming        99   
3    New Mexico State     12  New Mexico State            Nevada        99   
4         Idaho State     21  Nevada-Las Vegas       Idaho State        99   
..                ...    ...               ...               ...       ...   
644           Arizona     20         (12) Utah           Arizona        99   
645        New Mexico     10        Utah State        New Mexico        99   
646          Stanford     14          Stanford  Washington State        99   
647         Charlotte      7         Charlotte  Western Kentucky        99   
648          Maryland     10         Wisconsin          Maryland        99   

    away_rank  home_team_code  away_team_code  
0          99             136             117  
1          99             138             128  
2          99             145             289  
3          99             176             197  
4          99             174             154  
..        ...             ...             ...  
644        99              12              90  
645        99             215             200  
646        99             199             279  
647        99             126             284  
648        99             224             180  

[649 rows x 15 columns]

In [21]:
df1[df1['Loser'] == 'Arkansas']

Wk      Date      Time  Day                  Winner  Pts Unnamed: 6  \
406   7  8-Oct-22  12:00 PM  Sat  (23) Mississippi State   40        NaN   
613  11  5-Nov-22   4:00 PM  Sat            (23) Liberty   21          @   

        Loser  Pts.1               home_team          away home_rank  \
406  Arkansas     17  (23) Mississippi State      Arkansas        99   
613  Arkansas     19                Arkansas  (23) Liberty        99   

    away_rank  home_team_code  away_team_code  
406        99              71              92  
613        99             112              50

In [22]:
home_results = []
for index, row in df1.iterrows():
    if (row['home_team'] == row['Winner']):
        home_results.append(0)
    elif (row['home_team'] == row['Loser']):
        home_results.append(1)
    else:
        home_results.append(3)

In [23]:
df1.shape

(649, 15)

In [24]:
df1['home_results'] = home_results

In [25]:
df1.shape

(649, 16)

In [26]:
wins = df1['Winner']

In [27]:
winners = []
for i in wins:
    if (i[2] == ")"):
        winners.append(i[1:2])
        if(i[1] == "("):
            winners.append(i[1:3])
    else:
        winners.append(99)
    

In [28]:
len(winners)

649

In [29]:
df1['win_rank'] = winners

In [30]:
lose = df1['Loser']

In [31]:
losers = []
for i in lose:
    if (i[2] == ")"):
        losers.append(i[1:2])
        if(i[1] == "("):
            losers.append(i[1:3])
    else:
        losers.append(99)

In [32]:
len(losers)

649

In [33]:
df1['lose_rank'] = losers

In [34]:
df1 = df1.drop(columns=['Unnamed: 6'])

In [35]:
df1

Wk       Date      Time  Day            Winner  Pts             Loser  \
0     1  27-Aug-22   7:00 PM  Sat  Florida Atlantic   43         Charlotte   
1     1  27-Aug-22   5:00 PM  Sat     Florida State   47          Duquesne   
2     1  27-Aug-22   4:00 PM  Sat          Illinois   38           Wyoming   
3     1  27-Aug-22  10:00 PM  Sat            Nevada   23  New Mexico State   
4     1  27-Aug-22   3:30 PM  Sat  Nevada-Las Vegas   52       Idaho State   
..   ..        ...       ...  ...               ...  ...               ...   
644  11   5-Nov-22   7:30 PM  Sat         (12) Utah   45           Arizona   
645  11   5-Nov-22   3:30 PM  Sat        Utah State   27        New Mexico   
646  11   5-Nov-22   3:30 PM  Sat  Washington State   52          Stanford   
647  11   5-Nov-22  12:00 PM  Sat  Western Kentucky   59         Charlotte   
648  11   5-Nov-22  12:00 PM  Sat         Wisconsin   23          Maryland   

     Pts.1         home_team              away home_rank away_rank  \
0       13  Florida Atlantic         Charlotte        99        99   
1        7     Florida State          Duquesne        99        99   
2        6          Illinois           Wyoming        99        99   
3       12  New Mexico State            Nevada        99        99   
4       21  Nevada-Las Vegas       Idaho State        99        99   
..     ...               ...               ...       ...       ...   
644     20         (12) Utah           Arizona        99        99   
645     10        Utah State        New Mexico        99        99   
646     14          Stanford  Washington State        99        99   
647      7         Charlotte  Western Kentucky        99        99   
648     10         Wisconsin          Maryland        99        99   

     home_team_code  away_team_code  home_results win_rank lose_rank  
0               136             117             0       99        99  
1               138             128             0       99        99  
2               145             289             0       99        99  
3               176             197             1       99        99  
4               174             154             0       99        99  
..              ...             ...           ...      ...       ...  
644              12              90             0       99        99  
645             215             200             0       99        99  
646             199             279             1       99        99  
647             126             284             1       99        99  
648             224             180             0       99        99  

[649 rows x 17 columns]

In [36]:
df1 = df1.drop(columns=['home_team', 'away'])

In [37]:
df1

Wk       Date      Time  Day            Winner  Pts             Loser  \
0     1  27-Aug-22   7:00 PM  Sat  Florida Atlantic   43         Charlotte   
1     1  27-Aug-22   5:00 PM  Sat     Florida State   47          Duquesne   
2     1  27-Aug-22   4:00 PM  Sat          Illinois   38           Wyoming   
3     1  27-Aug-22  10:00 PM  Sat            Nevada   23  New Mexico State   
4     1  27-Aug-22   3:30 PM  Sat  Nevada-Las Vegas   52       Idaho State   
..   ..        ...       ...  ...               ...  ...               ...   
644  11   5-Nov-22   7:30 PM  Sat         (12) Utah   45           Arizona   
645  11   5-Nov-22   3:30 PM  Sat        Utah State   27        New Mexico   
646  11   5-Nov-22   3:30 PM  Sat  Washington State   52          Stanford   
647  11   5-Nov-22  12:00 PM  Sat  Western Kentucky   59         Charlotte   
648  11   5-Nov-22  12:00 PM  Sat         Wisconsin   23          Maryland   

     Pts.1 home_rank away_rank  home_team_code  away_team_code  home_results  \
0       13        99        99             136             117             0   
1        7        99        99             138             128             0   
2        6        99        99             145             289             0   
3       12        99        99             176             197             1   
4       21        99        99             174             154             0   
..     ...       ...       ...             ...             ...           ...   
644     20        99        99              12              90             0   
645     10        99        99             215             200             0   
646     14        99        99             199             279             1   
647      7        99        99             126             284             1   
648     10        99        99             224             180             0   

    win_rank lose_rank  
0         99        99  
1         99        99  
2         99        99  
3         99        99  
4         99        99  
..       ...       ...  
644       99        99  
645       99        99  
646       99        99  
647       99        99  
648       99        99  

[649 rows x 15 columns]

In [38]:
dates = df1["Date"]

In [39]:
len(dates)

649

In [40]:
month_num = []

for i in dates:
    if(i[1] =='-'):
        if(i[2:5] == 'Sep'):
            month_num.append('9')
        elif(i[2:5] == 'Oct'):
            month_num.append('10')
        elif(i[2:5] == 'Nov'):
            month_num.append('11')
        elif(i[2:5] == 'Aug'):
            month_num.append('8')
    elif(i[2] =='-'):
        if(i[3:6] == 'Aug'):
            month_num.append('8')
        elif(i[3:6] == 'Sep'):
            month_num.append('9')
        elif(i[3:6] == 'Oct'):
            month_num.append('10')
        elif(i[3:6] == 'Nov'):
            month_num.append('11')


In [41]:
len(month_num)

649

In [42]:
dates

0      27-Aug-22
1      27-Aug-22
2      27-Aug-22
3      27-Aug-22
4      27-Aug-22
         ...    
644     5-Nov-22
645     5-Nov-22
646     5-Nov-22
647     5-Nov-22
648     5-Nov-22
Name: Date, Length: 649, dtype: object

In [43]:
days = []

for i in dates: 
    if(i[1] == '-'):
        days.append(i[0])
    if(i[2] == "-"):
        days.append(i[0:2])

        


In [44]:
len(days)

649

In [45]:
years = []

for i in dates: 
    years.append('2022')
        
#         print(f'Hello {name}! This is {program}')

In [46]:
df1['months']= month_num
df1['years'] = years
df1['days'] = days

In [47]:
df1["dates"] = df1[
    'years'].astype(str) +"-"+ df1['months'].astype(str) + "-" + df1['days'].astype(str)

In [48]:
df1 = df1.drop(columns=['Date'])

In [49]:
df1 = df1.drop(columns=['months', 'years', 'days', 'Day'])

In [50]:
df1

Wk      Time            Winner  Pts             Loser  Pts.1 home_rank  \
0     1   7:00 PM  Florida Atlantic   43         Charlotte     13        99   
1     1   5:00 PM     Florida State   47          Duquesne      7        99   
2     1   4:00 PM          Illinois   38           Wyoming      6        99   
3     1  10:00 PM            Nevada   23  New Mexico State     12        99   
4     1   3:30 PM  Nevada-Las Vegas   52       Idaho State     21        99   
..   ..       ...               ...  ...               ...    ...       ...   
644  11   7:30 PM         (12) Utah   45           Arizona     20        99   
645  11   3:30 PM        Utah State   27        New Mexico     10        99   
646  11   3:30 PM  Washington State   52          Stanford     14        99   
647  11  12:00 PM  Western Kentucky   59         Charlotte      7        99   
648  11  12:00 PM         Wisconsin   23          Maryland     10        99   

    away_rank  home_team_code  away_team_code  home_results win_rank  \
0          99             136             117             0       99   
1          99             138             128             0       99   
2          99             145             289             0       99   
3          99             176             197             1       99   
4          99             174             154             0       99   
..        ...             ...             ...           ...      ...   
644        99              12              90             0       99   
645        99             215             200             0       99   
646        99             199             279             1       99   
647        99             126             284             1       99   
648        99             224             180             0       99   

    lose_rank      dates  
0          99  2022-8-27  
1          99  2022-8-27  
2          99  2022-8-27  
3          99  2022-8-27  
4          99  2022-8-27  
..        ...        ...  
644        99  2022-11-5  
645        99  2022-11-5  
646        99  2022-11-5  
647        99  2022-11-5  
648        99  2022-11-5  

[649 rows x 14 columns]

In [51]:
df1["dates"].value_counts()

2022-9-10     80
2022-9-17     69
2022-9-24     60
2022-9-3      56
2022-10-1     55
2022-10-8     52
2022-11-5     51
2022-10-22    49
2022-10-15    48
2022-10-29    42
2022-9-1      16
2022-8-27     11
2022-9-2       8
2022-9-30      5
2022-10-7      4
2022-9-22      3
2022-11-4      3
2022-10-27     3
2022-9-23      3
2022-9-16      2
2022-10-21     2
2022-9-29      2
2022-11-3      2
2022-11-2      2
2022-11-1      2
2022-10-28     2
2022-9-9       2
2022-9-4       2
2022-10-20     2
2022-10-14     2
2022-10-13     2
2022-10-2      1
2022-10-5      1
2022-9-18      1
2022-10-30     1
2022-9-5       1
2022-10-12     1
2022-10-19     1
Name: dates, dtype: int64

In [52]:
df1['dates'] = pd.to_datetime(df1['dates'])

In [53]:
df1["day_code"] = df1["dates"].dt.dayofweek

In [54]:
df1

Wk      Time            Winner  Pts             Loser  Pts.1 home_rank  \
0     1   7:00 PM  Florida Atlantic   43         Charlotte     13        99   
1     1   5:00 PM     Florida State   47          Duquesne      7        99   
2     1   4:00 PM          Illinois   38           Wyoming      6        99   
3     1  10:00 PM            Nevada   23  New Mexico State     12        99   
4     1   3:30 PM  Nevada-Las Vegas   52       Idaho State     21        99   
..   ..       ...               ...  ...               ...    ...       ...   
644  11   7:30 PM         (12) Utah   45           Arizona     20        99   
645  11   3:30 PM        Utah State   27        New Mexico     10        99   
646  11   3:30 PM  Washington State   52          Stanford     14        99   
647  11  12:00 PM  Western Kentucky   59         Charlotte      7        99   
648  11  12:00 PM         Wisconsin   23          Maryland     10        99   

    away_rank  home_team_code  away_team_code  home_results win_rank  \
0          99             136             117             0       99   
1          99             138             128             0       99   
2          99             145             289             0       99   
3          99             176             197             1       99   
4          99             174             154             0       99   
..        ...             ...             ...           ...      ...   
644        99              12              90             0       99   
645        99             215             200             0       99   
646        99             199             279             1       99   
647        99             126             284             1       99   
648        99             224             180             0       99   

    lose_rank      dates  day_code  
0          99 2022-08-27         5  
1          99 2022-08-27         5  
2          99 2022-08-27         5  
3          99 2022-08-27         5  
4          99 2022-08-27         5  
..        ...        ...       ...  
644        99 2022-11-05         5  
645        99 2022-11-05         5  
646        99 2022-11-05         5  
647        99 2022-11-05         5  
648        99 2022-11-05         5  

[649 rows x 15 columns]

In [55]:
df1.corr().loc[['home_results']]

Wk       Pts     Pts.1  home_team_code  away_team_code  \
home_results  0.13739 -0.127412  0.110057        0.138384        0.062202   

              home_results  day_code  
home_results           1.0  0.038636

In [56]:
df1.cov()

Wk         Pts       Pts.1  home_team_code  \
Wk               8.822467   -7.865505    4.058221        2.025840   
Pts             -7.865505  156.232162   21.533951      -54.082915   
Pts.1            4.058221   21.533951  108.797138       -1.803171   
home_team_code   2.025840  -54.082915   -1.803171     3124.223796   
away_team_code -25.725922  -14.897442 -106.522026      247.501201   
home_results     0.237740   -0.927783    0.668771        4.506185   
day_code        -0.000999    0.199650    0.052609       -1.342217   

                away_team_code  home_results  day_code  
Wk                  -25.725922      0.237740 -0.000999  
Pts                 -14.897442     -0.927783  0.199650  
Pts.1              -106.522026      0.668771  0.052609  
home_team_code      247.501201      4.506185 -1.342217  
away_team_code     5971.249282      2.800201 -1.308411  
home_results          2.800201      0.339392  0.013677  
day_code             -1.308411      0.013677  0.369239

In [57]:
df2 = df1.copy()

In [58]:
df2.head()

Wk      Time            Winner  Pts             Loser  Pts.1 home_rank  \
0   1   7:00 PM  Florida Atlantic   43         Charlotte     13        99   
1   1   5:00 PM     Florida State   47          Duquesne      7        99   
2   1   4:00 PM          Illinois   38           Wyoming      6        99   
3   1  10:00 PM            Nevada   23  New Mexico State     12        99   
4   1   3:30 PM  Nevada-Las Vegas   52       Idaho State     21        99   

  away_rank  home_team_code  away_team_code  home_results win_rank lose_rank  \
0        99             136             117             0       99        99   
1        99             138             128             0       99        99   
2        99             145             289             0       99        99   
3        99             176             197             1       99        99   
4        99             174             154             0       99        99   

       dates  day_code  
0 2022-08-27         5  
1 2022-08-27         5  
2 2022-08-27         5  
3 2022-08-27         5  
4 2022-08-27         5

In [59]:
df2 = df2.drop(columns=['Winner', 'Loser', 'Time', 'dates'])

In [60]:
df2

Wk  Pts  Pts.1 home_rank away_rank  home_team_code  away_team_code  \
0     1   43     13        99        99             136             117   
1     1   47      7        99        99             138             128   
2     1   38      6        99        99             145             289   
3     1   23     12        99        99             176             197   
4     1   52     21        99        99             174             154   
..   ..  ...    ...       ...       ...             ...             ...   
644  11   45     20        99        99              12              90   
645  11   27     10        99        99             215             200   
646  11   52     14        99        99             199             279   
647  11   59      7        99        99             126             284   
648  11   23     10        99        99             224             180   

     home_results win_rank lose_rank  day_code  
0               0       99        99         5  
1               0       99        99         5  
2               0       99        99         5  
3               1       99        99         5  
4               0       99        99         5  
..            ...      ...       ...       ...  
644             0       99        99         5  
645             0       99        99         5  
646             1       99        99         5  
647             1       99        99         5  
648             0       99        99         5  

[649 rows x 11 columns]

In [61]:
df3 = df2.drop(columns=['Pts', 'Pts.1', 'win_rank', 'lose_rank'])

In [62]:
df3

Wk home_rank away_rank  home_team_code  away_team_code  home_results  \
0     1        99        99             136             117             0   
1     1        99        99             138             128             0   
2     1        99        99             145             289             0   
3     1        99        99             176             197             1   
4     1        99        99             174             154             0   
..   ..       ...       ...             ...             ...           ...   
644  11        99        99              12              90             0   
645  11        99        99             215             200             0   
646  11        99        99             199             279             1   
647  11        99        99             126             284             1   
648  11        99        99             224             180             0   

     day_code  
0           5  
1           5  
2           5  
3           5  
4           5  
..        ...  
644         5  
645         5  
646         5  
647         5  
648         5  

[649 rows x 7 columns]

In [63]:
y = df3['home_results']
X = df3.drop(columns=['home_results'])

X_train, X_test, y_train, y_test = train_test_split(
     X, y, test_size=0.05, random_state=2)

In [64]:
X_train

Wk home_rank away_rank  home_team_code  away_team_code  day_code
350   6        99        99             162             253         5
192   4        99        99             172             291         5
109   3        99        99             136             241         5
22    2        99        99              40             281         3
161   3        99        99             208             201         5
..   ..       ...       ...             ...             ...       ...
534   9        99        99             209             208         5
584  10         3        99              84              31         5
493   9        99        99             114             176         5
527   9        99        99             195             258         5
168   3        99        99              18             246         5

[616 rows x 6 columns]

In [65]:
regr_1 = DecisionTreeRegressor(max_depth=4)

regr_2 = AdaBoostRegressor(
    DecisionTreeRegressor(max_depth=5), n_estimators=299, random_state=42
)

In [66]:
regr_1.fit(X_train, y_train)
regr_2.fit(X_train, y_train)

y_1 = regr_1.predict(X_train)
y_2 = regr_2.predict(X_train)

In [73]:
test_pred = regr_2.predict(X_test)


array([0.40277778, 0.51597052, 0.44219067, 0.46666667, 0.40277778,
       0.25      , 0.44219067, 0.        , 0.48178138, 0.21354167,
       0.25      , 0.51977401, 0.375     , 0.        , 0.50890585,
       0.43693694, 0.51977401, 0.45273632, 0.47413793, 0.44219067,
       0.50890585, 0.37974684, 0.25      , 0.47330097, 0.56375839,
       0.51977401, 0.56375839, 0.40277778, 0.45273632, 0.21354167,
       0.40277778, 0.47330097, 0.43693694])

In [68]:
print('RMSE: '+str(np.sqrt(mean_squared_error(y_test, test_pred))))
print('R Squared: '+str(r2_score(y_test, test_pred)))
print('MSE TEST: '+str(mean_squared_error(y_test, test_pred)))
print('MAE TEST: '+str(mean_absolute_error(y_test, test_pred)))
print('MAPE: '+str(mean_absolute_percentage_error(y_test, test_pred)))

RMSE: 0.4569672653828224
R Squared: -0.40372827096700226
MSE TEST: 0.20881908163145485
MAE TEST: 0.42987322941205286
MAPE: 1461928849953368.2


In [69]:
data = {'y_test': y_test, 'y_pred': y_train}
new_df = pd.DataFrame(data)
diff = (new_df['y_test'] - new_df['y_pred']).abs()
new_df['diff'] = diff
new_df = new_df.sort_values(by='diff', ascending=True)
new_df

y_test  y_pred  diff
0       NaN     0.0   NaN
1       0.0     NaN   NaN
2       NaN     0.0   NaN
3       NaN     1.0   NaN
4       NaN     0.0   NaN
..      ...     ...   ...
644     NaN     0.0   NaN
645     NaN     0.0   NaN
646     NaN     1.0   NaN
647     NaN     1.0   NaN
648     NaN     0.0   NaN

[649 rows x 3 columns]

In [70]:
df3.head()

Wk home_rank away_rank  home_team_code  away_team_code  home_results  \
0   1        99        99             136             117             0   
1   1        99        99             138             128             0   
2   1        99        99             145             289             0   
3   1        99        99             176             197             1   
4   1        99        99             174             154             0   

   day_code  
0         5  
1         5  
2         5  
3         5  
4         5

In [71]:
la = [[12, 99, 7, 112, 174, 5]]
lsu = pd.DataFrame(la, columns=['Wk', 'home_rank', 'away_rank', 'home_team_code', 'away_team_code', 'day_code'])
lsu

# arkansas home team code = 112
# lsu away team code = 174
# ga = [[11, 52, 1, 2, 1]]
# uga = pd.DataFrame(ga, columns=['Wk', 'Pts', 'home_results', 'win_rank', 'lose_rank'])
# uga

Wk  home_rank  away_rank  home_team_code  away_team_code  day_code
0  12         99          7             112             174         5

In [75]:
test_pred1 = regr_1.predict(lsu)
test_pred2 = regr_2.predict(lsu)

print(test_pred1)
print(test_pred2)

[0.47976879]
[0.56375839]


In [78]:
df[df['away_team_code'] == 174]

Wk       Date     Time  Day           Winner  Pts Unnamed: 6    Loser  \
347   6   1-Oct-22  7:00 PM  Sat  Louisiana State   21          @   Auburn   
458   8  15-Oct-22  7:00 PM  Sat  Louisiana State   45          @  Florida   

     Pts.1 home_team             away home_rank away_rank  home_team_code  \
347     17    Auburn  Louisiana State        99        99             115   
458     35   Florida  Louisiana State        99        99             135   

     away_team_code  
347             174  
458             174

In [79]:
df[df['home_team_code'] == 112]

Wk      Date     Time  Day        Winner  Pts Unnamed: 6     Loser  \
613  11  5-Nov-22  4:00 PM  Sat  (23) Liberty   21          @  Arkansas   

     Pts.1 home_team          away home_rank away_rank  home_team_code  \
613     19  Arkansas  (23) Liberty        99        99             112   

     away_team_code  
613              50

In [83]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.datasets import make_regression

In [84]:
regr = RandomForestRegressor(max_depth=2, random_state=0)
regr.fit(X_train, y_train)

RandomForestRegressor(max_depth=2, random_state=0)

In [86]:
print(regr.predict(lsu))


[0.42911376]


In [87]:
from sklearn.ensemble import AdaBoostRegressor
from sklearn.datasets import make_regression

In [88]:
regr = AdaBoostRegressor(random_state=0, n_estimators=100)
regr.fit(X_train, y_train)

AdaBoostRegressor(n_estimators=100, random_state=0)

In [89]:
AdaBoostRegressor(n_estimators=100, random_state=0)

AdaBoostRegressor(n_estimators=100, random_state=0)

In [92]:
pred_test = regr.predict(X_test)

In [91]:
regr.score(X_test, y_test)

-0.29421282757355294

In [93]:
data = {'y_test': y_test, 'y_pred': pred_test}
new_df = pd.DataFrame(data)
diff = (new_df['y_test'] - new_df['y_pred']).abs()
new_df['diff'] = diff
new_df = new_df.sort_values(by='diff', ascending=True)
new_df

y_test    y_pred      diff
80        0  0.113990  0.113990
55        0  0.113990  0.113990
209       0  0.113990  0.113990
141       0  0.113990  0.113990
89        0  0.113990  0.113990
1         0  0.113990  0.113990
165       0  0.113990  0.113990
344       0  0.182927  0.182927
452       0  0.182927  0.182927
639       0  0.182927  0.182927
158       0  0.313725  0.313725
30        0  0.313725  0.313725
205       0  0.313725  0.313725
515       0  0.478873  0.478873
569       0  0.478873  0.478873
249       0  0.478873  0.478873
559       0  0.478873  0.478873
637       0  0.478873  0.478873
333       0  0.478873  0.478873
556       0  0.478873  0.478873
294       0  0.478873  0.478873
508       0  0.478873  0.478873
275       0  0.478873  0.478873
395       0  0.478873  0.478873
397       0  0.478873  0.478873
440       0  0.478873  0.478873
321       0  0.478873  0.478873
459       1  0.478873  0.521127
467       1  0.478873  0.521127
581       1  0.478873  0.521127
248       1  0.478873  0.521127
176       1  0.113990  0.886010
37        1  0.113990  0.886010

In [94]:
regr.predict(lsu)

array([0.47887324])

## three predictions that lsu loses: .47976879, .42911376, .47887324
## two that lsu wins: .56375839, .61007972



In [95]:
from sklearn.datasets import make_regression
from sklearn.ensemble import GradientBoostingRegressor

In [96]:
reg = GradientBoostingRegressor(random_state=0)
reg.fit(X_train, y_train)

GradientBoostingRegressor(random_state=0)

In [98]:
pred_test = reg.predict(X_test)

In [99]:
data = {'y_test': y_test, 'y_pred': pred_test}
new_df = pd.DataFrame(data)
diff = (new_df['y_test'] - new_df['y_pred']).abs()
new_df['diff'] = diff
new_df = new_df.sort_values(by='diff', ascending=True)
new_df

y_test    y_pred      diff
89        0 -0.030695  0.030695
165       0  0.059006  0.059006
30        0  0.088650  0.088650
141       0  0.091527  0.091527
452       0 -0.091826  0.091826
80        0  0.114746  0.114746
55        0  0.120283  0.120283
1         0  0.153089  0.153089
639       0  0.211528  0.211528
205       0  0.213283  0.213283
294       0  0.225336  0.225336
275       0  0.242006  0.242006
321       0  0.265265  0.265265
508       0  0.272936  0.272936
249       0  0.289211  0.289211
209       0  0.311107  0.311107
333       0  0.322707  0.322707
395       0  0.404465  0.404465
397       0  0.404465  0.404465
158       0  0.426579  0.426579
556       0  0.447115  0.447115
440       0  0.454070  0.454070
344       0  0.455979  0.455979
581       1  0.527110  0.472890
569       0  0.499042  0.499042
515       0  0.556296  0.556296
459       1  0.404465  0.595535
559       0  0.605788  0.605788
248       1  0.337740  0.662260
637       0  0.678198  0.678198
467       1  0.294275  0.705725
176       1  0.201355  0.798645
37        1 -0.100377  1.100377

In [100]:
reg.predict(lsu)

array([0.61007972])